File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty()){
             return false;
        }

        int rows = matrix.size();
        int cols = matrix[0].size();

        int start = 0;
        int end = rows * cols - 1;

        while (start <= end) {
            int mid = (start + end) / 2;
            int midX = mid / cols;
            int midY = mid % cols;

            if (matrix[midX][midY] == target) {
                return true;
            } else if (matrix[midX][midY] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return false;
        




    }

  
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    if len(matrix)==0 || len(matrix[0]) == 0{
        return false
    }

    start:= 0
     var rows int = len(matrix)
    var cols int = len(matrix[0])

    end:= rows*cols -1

    for start<=end{
        mid:= (start+end)/2
        midX:= mid/cols
        midY:= mid%cols
        if matrix[midX][midY] == target{
            return true
        }else if  matrix[midX][midY] >target{
            end = mid-1;
        }else{
            start = mid+1;
        }
    }

    return false

}